### PR TITLE
RavenDB-18103 wait for document context to be updated

### PR DESCRIPTION
**Pull request by karmeli87**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18103

### CODE REVIEW

File: src/Raven.Server/Documents/Sharding/Handlers/Admin/ShardedAdminDatabasesHandler.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        public async Task WaitFor()
+        {
+            var database = GetStringQueryString("database");
+            var index = GetLongQueryString("index");
+
+            var result = Server.ServerStore.DatabasesLandlord.TryGetOrCreateDatabase(database);
+            if (result.DatabaseStatus != DatabasesLandlord.DatabaseSearchResult.Status.Sharded)
+                throw new NotSupportedException("This endpoint is only for sharded database");
+
+            await result.DatabaseContext.ShardedDatabaseContextIndexNotifications.WaitForIndexNotification(index, HttpContext.RequestAborted);
```
Comment (ayende) [inline src/Raven.Server/Documents/Sharding/Handlers/Admin/ShardedAdminDatabasesHandler.cs]: This is doing the wait on a _single_ node in the cluster. Is that your intent? Who is actually going to call this?

File: src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.cs [hunk 2]
```diff
@@ -76,17 +76,45 @@ public ShardedDatabaseContext(ServerStore serverStore, DatabaseRecord record)
...
+            ShardedDatabaseContextIndexNotifications.NotifyListenersAbout(index, e: null);
+        }
+
+        private int[] _uniqueShards;
+        public Memory<int> UniqueShards 
+        { 
+            get
+            {
+                if (_uniqueShards != null)
+                    return new Memory<int>(_uniqueShards);
```
Comment (ayende) [inline src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.cs]: Huh? Why not just the array directly?

File: src/Raven.Server/Documents/Commands/WaitForRaftCommands.cs [hunk 3]
```diff
@@ -2,27 +2,30 @@
...
 
 namespace Raven.Server.Documents.Commands
 {
     public class WaitForRaftCommands : RavenCommand
     {
         private readonly List<long> _indexes;
+        private readonly int? _shardNumber;
 
-        public WaitForRaftCommands(List<long> indexes)
+        public WaitForRaftCommands(List<long> indexes, int? shardNumber = null)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Commands/WaitForRaftCommands.cs]: If we need to do have it then it means that we are using wrong executor at some point

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedRachisDatabaseHandler.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+                var blittable = await context.ReadForMemoryAsync(RequestBodyStream(), "raft-index-ids");
+                var commands = JsonDeserializationServer.WaitForRaftCommands(blittable);
+
+                foreach (var index in commands.RaftCommandIndexes)
+                {
+                    await DatabaseContext.ShardedDatabaseContextIndexNotifications.WaitForIndexNotification(index, HttpContext.RequestAborted);
+
+                    for (int i = 0; i < DatabaseContext.ShardCount; i++)
+                    {
+                        var name = ShardHelper.ToShardName(DatabaseContext.DatabaseName, i);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedRachisDatabaseHandler.cs]: DatabaseContext should know which shards reside on that node? Can we use DatabaseRecord to calculate this?

File: src/Raven.Server/Documents/Sharding/Handlers/ShardedRachisDatabaseHandler.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+                        if (Server.ServerStore.DatabasesLandlord.DatabasesCache.TryGetValue(name, out var task))
+                        {
+                            try
+                            {
+                                var physicalDatabase = await task;
+                                await physicalDatabase.RachisLogIndexNotifications.WaitForIndexNotification(index, HttpContext.RequestAborted);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/ShardedRachisDatabaseHandler.cs]: You are also 'eating' timeouts, etc. This is not what we want?
Do try/catch only on 
```
var physicalDatabase = await task;
```

File: src/Raven.Server/Documents/Sharding/Operations/WaitForDatabaseContextUpdateOperation.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Collections.Generic;
+using Raven.Client.Http;
+using Raven.Server.Documents.Commands;
+
+namespace Raven.Server.Documents.Sharding.Operations
+{
+    public readonly struct WaitForDatabaseContextUpdateOperation : IShardedOperation
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Operations/WaitForDatabaseContextUpdateOperation.cs]: Isn't this an overkill?

Comment (ppekrol) [reply src/Raven.Server/Documents/Sharding/Operations/WaitForDatabaseContextUpdateOperation.cs]: Ping me tomorrow.

File: src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.cs [hunk 7]
```diff
@@ -45,6 +45,8 @@ public partial class ShardedDatabaseContext : IDisposable
 
         public int[] FullRange;
 
+        public RachisLogIndexNotifications ShardedDatabaseContextIndexNotifications;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.cs]: do not like the name, move back to `RachisLogIndexNotifications`

File: src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.cs [hunk 8]
```diff
@@ -147,6 +154,15 @@ public void Dispose()
                 }
             }
 
+            try
+            {
+                AllNodesExecutor.Dispose();
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.cs:161]: Write a TODO here that this needs an ExceptionAggregator like DocumentDatabase

**Review by ppekrol — APPROVED** [review]

**Pull request merged**